### PR TITLE
use background process to survive popup auto-closing (#20)

### PR DESCRIPTION
**Pull request by zsau**
Initial stab at fixing #20. Thoughts?

### CODE REVIEW

**Comment by martinhoefling** [comment]
Thx @zsau, I‘ll look through it as soon as I have some more time. It definitely requires some tests as I try to get the codebase mostly covered at least by unit tests. Maybe @pharb can take a look at it already.

File: manifests/firefox-manifest.json [hunk 1]
```diff
@@ -35,6 +35,13 @@
     }
   ],
 
+  "background": {
+    "scripts": [
```
Comment (Pharb) [inline manifests/firefox-manifest.json]: We should play safe and always include `vendor/browser-polyfill.js` before loading other scripts.

**Comment by zsau** [comment]
Thanks for the feedback, @Pharb.

Re: error handling, I mostly just tried to use what was already there. I'm not sure about the best method for reporting errors if the popup is already closed, though.

Re: coverage, I noticed that the coverage analysis doesn't even "see" background.js, so at least that should probably be added to whatever appropriate build file.

**Comment by Pharb** [comment]
@zsau Thanks for your quick updates 👍 

About error handling:
- While the popup is open the best way is probably to just return a Promise and reject it. I tried this in Firefox and it seemed to work fine with these changes: https://github.com/Pharb/gopassbridge/commit/3725cd89798311668af71b9b8747460fc43eabfa
- Showing an error message when the popup is closed is probably a bit harder, because we might have to inject some error message into the content page if we want to do that. Personally I don't consider this a high priority, it might be enough for now to just open a separate issue for this to keep it in mind for the future

About the test coverage:
- JS files are currently only included in the coverage report when they are actually referenced by a test file. This means that the minimum to include it would be to create `tests/unit/background.test.js` with at least something like this:
```
'use strict';
require('background.js');
describe('background script', function() {
    test('works', function() {});
});
```
- Even better if there were actual tests, but this can also be done in #45 😉

**Comment by martinhoefling** [comment]
I tried it out and tried to fix copying to the clipboard, but this does not seem to be possible from within a background script (https://stackoverflow.com/questions/45622608/copy-text-to-clipboard-from-background-script-in-a-firefox-webextension). How about reverting that part and focus on filling the credentials only?

**Comment by zsau** [comment]
I just had to bounce back to the content script to get copy-to-clipboard working.

As for error handling, how do you like this approach? Caveat: I found out while debugging that gopass isn't returning useful status codes (https://github.com/gopasspw/gopass/issues/834), so until that's fixed we still won't display anything in many failure scenarios.

**Comment by martinhoefling** [comment]
Thx for the additions. I would prefer not to leak any information to the main content as it is also accessible from javascript running in the site context. Otherwise, we would need to filter the error messages coming from gopass, which I would like to avoid. So far, you can disable the style modifications in the settings and thus only fill in the values (no gopass usage revealed). 

How about returning the error to the popup **and** logging to the console. As such, normal users will see the error and in your case, you can still access the error messages from the console.

**Comment by zsau** [comment]
Fair point. Would using an iframe address the information leakage? I'd like to avoid any reduced functionality when pinentry appears, including error reporting.

**Comment by martinhoefling** [comment]
Hmm, I don't think that an iframe is properly sandboxing the output. I suggest we first try out if opening a new window is an option. Might be a bit ugly but should do the trick. This should be doable for the copied message and error messages. For the latter, we can do this by default, in the "copied" message case, I would prefer to leave it as an option for pinentry lovers. @Pharb do you have any other ideas how to solve this properly?

**Comment by Pharb** [comment]
These are the options that came to my mind that might work to get an error shown to the user if the popup is closed:
- alert
  - simple to maintain
  - ugly in design, but probably sufficient for this case
  - security: might be a problem because `window.alert` can be replaced and thus leak some information to the content script
- notification
  - requires extra permission, which could also be optional? Maybe a user setting.
  - gopass itself uses desktop notifications sometimes
  - security: should be fine?
- main content
  - very flexible what to do, but IMHO gopassbridge should try to be as non-invasive as possible with the content itself and to keep the interactions with the content minimal to avoid security bugs
  - security: leaks into content
- new tab/window
  - for UX IMHO as ugly as calling alert(), but probably ok if it is turned off by default as user setting?
  - probably some effort to maintain
  - security: should be fine

**Comment by zsau** [comment]
AFAIK, cross-domain iframes are supposed to be completely isolated from the parent, though I'm not sure how to make the browser treat an addon-generated iframe as cross-domain.

But notifications sound like a better option to me, anyway.

**Comment by martinhoefling** [comment]
concerning security: new tab/window risks that other extensions gain access to the clear text password. So I would go for notifications if no unexpected problems arise, since they support both use cases (with and without pinentry) sufficiently. I don't mind the extra permission.

**Comment by martinhoefling** [comment]
Well, notifications cannot solve to copy to clipboard issue right? We could do that directly from gopass. What do you think?

**Comment by Pharb** [comment]
@martinhoefling Moving clipboard handling back to gopass sounds very good.

As this PR is getting a bit out of scope of the original issue I would like to split the following separate parts:

1.  Add the ability to trigger notifications from gopassbridge #48
2. Add clipboard handling to gopass jsonapi and use it #49

**Comment by martinhoefling** [comment]
We still need a solution for the detailed view. We should also be able to copy to clipboard from there.
How about just storing the last entry in the and use it from cache? This would require reopening the popup for Pinentry users.

**Comment by Pharb** [comment]
Now that #48 is merged, the next steps might be:
1. replace status error message with notification when popup is closed in this PR
2. move clipboard handling back to popup script in this PR
3. check that the basics work in Firefox and Chrome with this PR
4. get this PR merged
5. fix clipboard for the pinentry use case separately in #49

**Comment by martinhoefling** [comment]
I rebased the commits and refactored the background script a bit to make it future-proof (more messages). The login case should work now. We still need some tests / fixes but if complete, we can merge this I think. From there we can go ahead as @Pharb  suggested. @Pharb and @zsau, what are your thoughts?

**Comment by martinhoefling** [comment]
Thx @zsau and @Pharb

**Comment by zsau** [comment]
Awesome. Thanks for writing the tests and for all the other help, @martinhoefling.

**Pull request merged**